aptos-move/move-examples/vote_lockup/sources/vote.move

```diff
@@ -0,0 +1,389 @@
+    }
+
+    /// Withdraw the $VOTING from a voting certificate. The certificate must have expired.
+    public entry fun withdraw_entry(
+        owner: &signer,
+        certificate: Object<VotingCertificate>,
+    ) acquires VotingCertificate {
+        let assets = withdraw(owner, certificate);
+        primary_fungible_store::deposit(signer::address_of(owner), assets);
+    }
+
+    /// Withdraw the $VOTING from a voting certificate. The certificate must have expired.
+    public fun withdraw(
+        owner: &signer,
+        certificate: Object<VotingCertificate>,
+    ): FungibleAsset acquires VotingCertificate {
+        // Extract the unlocked $VOTING and burn the ve token.
+        let tokens = voting_token::withdraw(certificate, fungible_asset::balance(certificate));
+        let end_epoch = owner_only_destroy_certificate(owner, certificate);
+        // This would fail if the lockup has not expired yet.
+        assert!(end_epoch <= epoch::now(), ELOCKUP_HAS_NOT_EXPIRED);
+        tokens
+        // Withdraw doesn't need to update total voting power because this lockup should not have any effect on any
+        // epochs, including the current one, as it has already expired.
+    }
+
+    fun update_manifested_total_supply(
+        old_amount: u64,
+        old_lockup_end_epoch: u64,
+        new_amount: u64,
+        new_lockup_end_epoch: u64,
+    ) acquires VoteConfig {
+        assert!(
+            new_amount > old_amount || new_lockup_end_epoch > old_lockup_end_epoch,
+            EINVALID_LOCKUP_CHANGE,
+        );
+
+        // We only need to update the total supply starting from the current epoch since the total voting powers of
+        // past epochs are already set in stone.
+        let curr_epoch = epoch::now();
+        let total_voting_power_per_epoch = &mut unchecked_vote_config().unscaled_total_voting_power_per_epoch;
+        while (curr_epoch < new_lockup_end_epoch) {
+            // Old epoch delta can be zero if there was no previous lockup (old_amount = 0) or lockup has expired.
+            let old_epoch_delta = if (old_amount == 0 || old_lockup_end_epoch <= curr_epoch) {
+                0
+            } else {
+                old_amount * (old_lockup_end_epoch - curr_epoch)
+            };
+            let new_epoch_delta = new_amount * (new_lockup_end_epoch - curr_epoch);
+            // This cannot underflow due to the assertion that either the amount or the lockup duration or both must
+            // increase.
+            let voting_power_delta = ((new_epoch_delta - old_epoch_delta) as u128);
+            if (smart_table::contains(total_voting_power_per_epoch, curr_epoch)) {
+                let total_voting_power = smart_table::borrow_mut(total_voting_power_per_epoch, curr_epoch);
+                *total_voting_power = *total_voting_power + voting_power_delta;
+            } else {
+                smart_table::add(total_voting_power_per_epoch, curr_epoch, voting_power_delta);
+            };
```

Comment:
could minimize to borrow_mut_with_default of 0 followed by the update to total power.

Reply:
Good call. Updated.